### PR TITLE
chore: refactor integration test codegen

### LINKED ISSUE #2337: Remove duplication in integration test codegen
### Problem

These functions are currently almost identical except for the subdirectory and the snippet which is being written

https://github.com/noir-lang/noir/blob/fd291971f2060438c4b529e4cb5b5911a9939418/crates/nargo_cli/build.rs#L42-L44

### Happy Case

We should abstract these functions so that we can have a single implementation which receives a subdirectory and template and fills it in for all the test cases.

### Alternatives Considered

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

No

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by jtfirek**
# Description

Refactors the integration test codegen

## Problem

This file had duplicate code

Resolves #2337

## Summary

I pass the template in as a closure into a single function that handles all of the cases

## Documentation

no documentation needed 

# PR Checklist

- [✅ ] I have tested the changes locally.
- [✅] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings. 

I am testing using `cd crates/nargo_cli && cargo tests`

I have set up my dev environment using `allow direnv`, but it looks like I am failing a good bit of the tests. Do you think their might be something wrong with my development environment or are my changes causing these failures? 
Screenshot 2023-09-04 at 3 04 42 PM

It seems like they are failing because because of this error:
Screenshot 2023-09-04 at 3 06 26 PM

### CODE REVIEW

File: crates/nargo_cli/build.rs [hunk 1]
```diff
@@ -119,51 +84,81 @@ fn compile_success_empty_{test_name}() {{
...
+        "compile_success_empty",
+        &*generate_compile_success_empty_code,
+    );
+    generate_tests(
+        &mut test_file,
+        &test_dir,
+        "compile_success_contract",
+        &*generate_compile_success_contract_code,
+    );
+    generate_tests(&mut test_file, &test_dir, "compile_failure", &*generate_compile_failure_code);
```
Comment (TomAFrench) [inline crates/nargo_cli/build.rs:153]:
```suggestion
    generate_tests(
        &mut test_file,
        &test_dir,
        "execution_success",
        &generate_execution_success_code,
    );
    generate_tests(
        &mut test_file,
        &test_dir,
        "compile_success_empty",
        &generate_compile_success_empty_code,
    );
    generate_tests(
        &mut test_file,
        &test_dir,
        "compile_success_contract",
        &generate_compile_success_contract_code,
    );
    generate_tests(&mut test_file, &test_dir, "compile_failure", &generate_compile_failure_code);
```

These don't need to be dereferenced

**Review by TomAFrench — COMMENTED** [review]
One small nit but otherwise looks good.

Thanks for helping with this!

File: crates/nargo_cli/build.rs [hunk 2]
```diff
@@ -119,51 +84,81 @@ fn compile_success_empty_{test_name}() {{
...
 }
 
-fn generate_compile_failure_tests(test_file: &mut File, test_data_dir: &Path) {
-    let test_sub_dir = "compile_failure";
+fn generate_tests(
+    test_file: &mut File,
+    test_data_dir: &Path,
+    test_sub_dir: &str,
+    generate_code: &TestCodeGenerator,
```
Comment (jfecher) [inline crates/nargo_cli/build.rs:160]: We don't need a `Box<dyn Fn...>` here if we're just taking it in as an argument. `impl Fn` without boxing should work as well.

File: crates/nargo_cli/build.rs [hunk 3]
```diff
@@ -119,51 +84,81 @@ fn compile_success_empty_{test_name}() {{
...
-
-fn generate_compile_success_contract_tests(test_file: &mut File, test_data_dir: &Path) {
-    let test_sub_dir = "compile_success_contract";
-    let test_data_dir = test_data_dir.join(test_sub_dir);
-
-    let test_case_dirs =
-        fs::read_dir(test_data_dir).unwrap().flatten().filter(|c| c.path().is_dir());
+                test_dir = test_dir.display(),
+            )
+            .expect("Could not write templated test file.");
```
Comment (jfecher) [inline crates/nargo_cli/build.rs:89]: It looks like the only parts that differ between each of these generator functions is the raw string that is concatenated in each, is this correct? If so, we can forego the functions altogether and just pass in the strings directly into `generate_tests`.

Comment (jtfirek) [reply crates/nargo_cli/build.rs]: Good point! I removed the functions and I am now just passing the strings directly in!

File: crates/nargo_cli/build.rs (hunk 2 above)
Comment (jtfirek) [reply crates/nargo_cli/build.rs]: Fixed I am just passing a string in now!

File: crates/nargo_cli/build.rs (hunk 1 above)
Comment (jtfirek) [reply crates/nargo_cli/build.rs]: dang sorry about missing this. Fixed it!

**Comment by jtfirek** [comment]
@jfecher and @TomAFrench any updates on this?

**Comment by TomAFrench** [comment]
Apologies for the lack of response. This code has since been refactored elsewhere so I'm going to close this PR.

**Pull request closed without merging**